src/main/java/BuildingsCatchingWater.java

```diff
@@ -0,0 +1,58 @@
+/**
+ * Given an array of integers (that represents buildings in sequential order), compute how much rain water
+ * they would catch. The amount of water 'caught' for a given building can be computed by checking the
+ * bounding buildings on either side.
+ */
+public class BuildingsCatchingWater {
+
+
+    public static int compute(int[] buildings) throws IllegalArgumentException {
+
+        if(buildings == null) {
+            throw new IllegalArgumentException("invalid input array of buildings");
+        }
+        if(buildings.length < 3) {
+            return 0;
+        }
+
+        int left = 0;
+        int right = buildings.length - 1;
+
+        // left and right trim
+        while(left < right && buildings[left] <= buildings[left+1]) {
+            left++;
+        }
+        while(left < right && buildings[right - 1] >= buildings[right]) {
+            right--;
+        }
+        if(left == right) {
+            return 0;
+        }
+        return BCW(buildings, left, right);
+    }
+
+    private static int BCW(int[] buildings, int left, int right) {
```

Comment:
this is a good recursive implementation, but it's going to get expensive if larger arrays are passed in. can we do this incrementally, without recursion?